src/javascript/url-shortener.js

```diff
@@ -12,71 +12,327 @@
+      ) {
+        const data = event.data;
+        if (data.error) {
+          reject(new Error(data.error));
+        } else {
+          resolve(data);
+        }
+      }
+    };
+    window.addEventListener('message', messageListener, {once: true});
+    cleanup(() => window.removeEventListener('message', messageListener));
+
+    const authUrl = 'https://bitly.com/oauth/authorize' + encodeQuery({
+      client_id: bitlyClientId,
+      // The auth callback does about half the work here: it receieves an
+      // authorization code, converts it into a token, then sends the token
+      // back to us. This happens server side because it requires the
+      // client_secret, which we don't want to share with our users.
+      redirect_uri: window.location.origin + '/bitly-api-token-handler/',
+    });
+    // Open a new window to do authentication.
+    const childWindow = window.open(authUrl, '_blank');
+    if (childWindow == null) {
+      throw new Error(
+        'Failed open new window for authorizing bitly. ' +
+        'Are you blocking popups?'
+      );
+    }
+
+    // Close the window as soon as auth flow completes.
+    cleanup(() => childWindow.close());
+
+    // If the user closes the window, that's an error
+    const intervalHandle = window.setInterval(() => {
+      if (childWindow.closed) {
+        reject(new Error(
+          'Authorization window closed before authorization was completed'
+        ));
+      }
+    }, 1000);
+    cleanup(() => window.clearInterval(intervalHandle));
+
+    // If the user takes too long, that's an error.
+    const timeoutHandle = window.setTimeout(() => {
+      reject(new Error('Timed out waiting for authorization'));
+    }, 1000 * 60 * 5);
+    cleanup(() => window.clearTimeout(timeoutHandle));
   });
-}
+
+  // We have a token! Save it to localStorage, then retry the API call. At this
+  // point, all errors should just be sent to the client; we've done all we can
+  // to make it successful
+
+  setToken(token);
+
+  return await createBitlink({
+    longUrl: longUrl,
+    token: token,
+    checkForbidden: false,
+  });
+});
+
+// Generic bitly v4 API calls. Makes a request to
+// https://api-ssl.bitly.com/v4/{endpoint}, using the options. If payload is
+// given, it is JSON encoded and used as the body, and the content-type header
+// is set. Uses token for authentication. If a payload is given and no method
+// is set, the method is POST. This function uses fetch api, so all options
+// are as to the fetch() function.
+//
+// HTTP errors are rejected with a summary message; otherwise, the response
+// JSON payload is returned. If checkForbidden is true, 403 errors cause the
+// forbiddenError object to be returned (as a rejection); this allows for
+// easily detecting this case for OAuth flow.
+const bitlyApiFetch = async ({
+  token,
+  endpoint,
+  options,
+  payload=undefined,
+  checkForbidden=false,
+}) => {
+  options = options || {};
```

Comment:
probably would be better to do this as default arguments like you're doing with payload and checkForbidden.

Reply:
Is it safe to do that? I assumed that Javascript had the same "mutable defaults" problem that Python does.

Reply:
Though I guess since I'm not mutating it, it shouldn't be a problem anyway.